game-headed/src/main/java/org/triplea/game/client/ui/javafx/screens/MapSelection.java

```diff
@@ -0,0 +1,149 @@
+import javafx.scene.layout.TilePane;
+import javafx.scene.web.WebView;
+import lombok.AccessLevel;
+import lombok.Setter;
+import org.triplea.game.client.parser.DefaultGameDetector;
+import org.triplea.game.client.parser.GameDetector;
+import org.triplea.game.client.ui.javafx.screen.ControlledScreen;
+import org.triplea.game.client.ui.javafx.screen.ScreenController;
+import org.triplea.game.client.ui.javafx.util.FxmlManager;
+import org.triplea.util.LocalizeHtml;
+
+public class MapSelection implements ControlledScreen<ScreenController<FxmlManager>> {
+
+  private final GameDetector gameDetector;
+  private final BiFunction<String, String, String> linkLocalizer;
+
+  @Setter(
+      value = AccessLevel.PACKAGE,
+      onMethod_ = {@VisibleForTesting})
+  @FXML
+  private Node root;
+
+  @FXML private TilePane mapContainer;
+
+  @Setter(
+      value = AccessLevel.PACKAGE,
+      onMethod_ = {@VisibleForTesting})
+  @FXML
+  private Node previewContainer;
+
+  @Setter(
+      value = AccessLevel.PACKAGE,
+      onMethod_ = {@VisibleForTesting})
+  @FXML
+  private WebView previewWindow;
+
+  @FXML private Button selectButton;
+  @FXML private Button detailsButton;
+
+  @Setter(
+      value = AccessLevel.PACKAGE,
+      onMethod_ = {@VisibleForTesting})
+  private GameChooserEntry selectedGame;
+
+  private boolean loaded = false;
+
+  private ScreenController<FxmlManager> screenController;
+
+  // Constructor used via FXML reflection
+  // to initialize this controller.
+  @SuppressWarnings("unused")
+  public MapSelection() {
+    this(new DefaultGameDetector(), LocalizeHtml::localizeImgLinksInHtml);
+  }
+
+  @VisibleForTesting
+  MapSelection(
+      final GameDetector gameDetector, final BiFunction<String, String, String> linkLocalizer) {
+    this.gameDetector = gameDetector;
+    this.linkLocalizer = linkLocalizer;
+  }
+
+  private Node createMapListing(final GameChooserEntry gameChooserEntry) {
+    final var button = new Button(gameChooserEntry.getGameName());
+    // Placeholder image
+    final var imageView =
+        new ImageView(new Image("https://triplea-game.org/images/missing_map.png", true));
+    imageView.setPreserveRatio(true);
+    imageView.setFitHeight(150);
+    imageView.setFitWidth(170);
+    button.setGraphic(imageView);
+    button.setWrapText(true);
+    button.setPrefSize(200, 200);
+    button.setContentDisplay(ContentDisplay.TOP);
+    button.setOnAction(
+        event -> {
+          selectedGame = gameChooserEntry;
+          selectButton.setDisable(false);
+          detailsButton.setDisable(false);
+          mapContainer.getChildren().forEach(node -> node.getStyleClass().remove("selected"));
+          button.getStyleClass().add("selected");
+        });
+    return button;
+  }
+
+  @Override
+  public void connect(final ScreenController<FxmlManager> screenController) {
+    this.screenController = screenController;
+  }
+
+  @Override
+  public void onShow() {
+    if (loaded) {
```

Comment:
When is this class instantiated? Can 'onShow' be moved to constructor, any problems with that?

Reply:
I think I might need to explain the pattern to you (or it might be better if you tried to build a simple screen on your own/polish an existing one, similar to my ansible experience).
Basically the *Controller classes are instantiated by JavaFX once for every FXML file, so controllers are essentially singletons, somewhat similar to dependency injection in spring.

`onShow()` was introduced because when I started working on this I had a severe freeze on startup, which I thought would come from the heavy game loading.
`onShow()` is only called when a screen is switched to/made visible on the stack, so this way the freeze would only occur once when people would actually try to play a game, which felt way better.

In the end it turned out that the freeze was because all of the thumbnails were loaded synchronously on the JavaFX thread, so once I changed this addition is not really needed anymore.
In fact if we moved the "loading" mechanism somewhere else, we could have the maps load in the background, so that the user might not even notice that anything took 3 seconds to load because everything is already loaded when they actually start a game.
So thanks for reminding me.

This class is initialized on startup, along with all the other FXML controllers, however at this point none of the dependency injection has occured.
What you want to do is add a method like this
```java
@FXML
void initialize() {
  // initialize everything here
}
```
which is called by the FXML loader once the Nodes have been injected.